aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
 
   OperatorName op_name = schema.operator_name();
 
-  auto op = findOrRegisterSchema_(std::move(schema));
+  const auto found = findSchema(op_name);
+  OperatorHandle op = [&] {
+    if (found == c10::nullopt) {
+      // Initial def
+      operators_.emplace_back(std::move(schema));
+      OperatorHandle op(--operators_.end());
+      operatorLookupTable_.write([&] (ska::flat_hash_map<OperatorName, OperatorHandle>& operatorLookupTable) {
+        operatorLookupTable.emplace(op_name, op);
+      });
+      // Call the hook after the operator is setup!
+      listeners_->callOnOperatorRegistered(op);
+      return op;
+    } else {
+      // Re-registration for BC reasons, test for consistency
+      OperatorHandle op(*found);
+      TORCH_INTERNAL_ASSERT(op.operatorIterator_->def_and_impl_count >= 1);
+      // NB: the operator may have been impl'ed but not def'ed
+      if (op.operatorIterator_->def_count == 0) {
+        op.operatorIterator_->op.registerSchema(std::move(schema));
```

Comment:
I think you have to call the listener here too, listener should trigger at the first def() call even if there's already an impl. Maybe it's better to keep the listener handling outside of this if/else, similar to how it was before. Also make sure to add a test case for this!

Reply:
@smessmer that's how I'm reading the code as well - we wind up here if `registerImpl` is called before `registerDef` on this op.

Reply:
Yes this is a bug.

Reply:
Fixed!